addons/controllers/clusterbootstrap_controller.go

```diff
@@ -606,26 +952,47 @@ func (r *ClusterBootstrapReconciler) updateValuesFromProvider(cluster *clusterap
 		providerLabels := newProvider.GetLabels()
 		if providerLabels == nil {
 			newProvider.SetLabels(map[string]string{
-				types.PackageNameLabel: pkg.RefName,
+				// A package refName could contain characters that K8S does not like as a label value.
+				// For example, kapp-controller.tanzu.vmware.com.0.30.0+vmware.1-tkg.1 is a
+				// valid package refName, but it contains "+" that K8S complains. We parse the refName by replacing
+				// + to ---.
+				types.PackageNameLabel: util.ParseStringForLabel(pkg.RefName),
 				types.ClusterNameLabel: cluster.Name,
 			})
 		} else {
-			providerLabels[types.PackageNameLabel] = pkg.RefName
+			providerLabels[types.PackageNameLabel] = util.ParseStringForLabel(pkg.RefName)
 			providerLabels[types.ClusterNameLabel] = cluster.Name
 		}
 
-		newProvider.SetName(fmt.Sprintf("%s-%s-package", cluster.Name, packageShortName(pkg.RefName)))
-		log.Info("cloning provider", "provider", newProvider)
-		newProvider, err = r.dynamicClient.Resource(*gvr).Namespace(bootstrap.Namespace).Create(r.context, newProvider, metav1.CreateOptions{})
+		newProvider.SetName(fmt.Sprintf("%s-%s-package", cluster.Name, pkgRefName))
+		newProvider.SetNamespace(bootstrap.Namespace)
+		log.Info(fmt.Sprintf("cloning provider %s/%s to namespace %s", cbTemplateNamespace, newProvider.GetName(), bootstrap.Namespace), "gvr", gvr)
+		// newProvider and createdOrUpdatedProvider are different. The newProvider is the one we want apiserver to accept,
```

Comment:
this comment makes the creation seem a lot more complex than it actually is

Reply:
making this change and adding that comment is because there is a bug on the original code when there are errors on create.
```
newProvider, err = r.dynamicClient.Resource(*gvr).Namespace(bootstrap.Namespace).Create(r.context, newProvider, metav1.CreateOptions{})
```

`newProvider` is nil and the code returns it. `ensureOwnerRef` uses that nil object to update ownerRef which causes NPE.

We might need a follow up to refactor the code a little bit.